Terminal.Gui/Views/Menu/MenuBar.cs

```diff
@@ -0,0 +1,1392 @@
+	/// If the menu is open, it will close the menu bar.
+	/// </summary>
+	/// <returns></returns>
+	bool SelectOrRun ()
+	{
+		if (!IsInitialized || !Visible) {
+			return true;
+		}
+
+		if (_menuBarItemToActivate != -1) {
+			Activate (_menuBarItemToActivate);
+		} else if (_menuItemToSelect != null) {
+			Run (_menuItemToSelect.Action);
+		} else {
+			if (IsMenuOpen && _openMenu != null) {
+				CloseAllMenus ();
+			} else {
+				OpenMenu ();
+			}
+
+		}
+		return true;
+	}
+
+	/// <inheritdoc/>
+	public override bool OnInvokeKeyBindings (KeyEventArgs keyEvent)
+	{
+		// This is a bit of a hack. We want to handle the key bindings for menu bar but
+		// InvokeKeyBindings doesn't pass any context so we can't tell which item it is for.
+		// So before we call the base class we set SelectedItem appropriately.
+		// TODO: Figure out if there's a way to have KeyBindings pass context instead. Maybe a KeyBindingContext property?
+
+		var key = keyEvent.Key;
+
+		if (ContainsKeyBinding (key)) {
+			_menuBarItemToActivate = -1;
+			_menuItemToSelect = null;
+
+			// Search for shortcuts first. If there's a shortcut, we don't want to activate the menu item.
+			for (var i = 0; i < Menus.Length; i++) {
+				// Recurse through the menu to find one with the shortcut.
+				if (FindShortcutInChildMenu (key, Menus [i], out _menuItemToSelect)) {
+					_menuBarItemToActivate = i;
+					return base.OnInvokeKeyBindings (keyEvent);
+				}
+
+				// Now see if any of the menu bar items have a hot key that matches
+				// Technically this is not possible because menu bar items don't have 
+				// shortcuts or Actions. But it's here for completeness. 
+				var shortcut = Menus [i]?.Shortcut;
+				if (key == shortcut) {
+					throw new InvalidOperationException ("Menu bar items cannot have shortcuts");
+					//_menuBarItemToActivate = -1;
+					//_menuItemToActivate = Menus [i];
+					//return base.OnInvokeKeyBindings (keyEvent);
+				}
+
+			}
+
+			// Search for hot keys next.
+			for (var i = 0; i < Menus.Length; i++) {
+				if (IsMenuOpen) {
+					// We don't need to do anything because `Menu` will handle the key binding.
+				}
+
+				// No submenu item matched (or the menu is closed)
+
+				// Check if one of the menu bar item has a hot key that matches
+				var hotKeyValue = Menus [i]?.HotKey.Value ?? default;
+				var hotKey = (Key)hotKeyValue;
+				if (hotKey != Key.Null) {
+					var matches = key == hotKey || key == (hotKey | Key.AltMask);
+					if (IsMenuOpen) {
+						// If the menu is open, only match if Alt is not pressed.
+						matches = key == hotKey;
+					}
+
+					if (matches) {
+						_menuBarItemToActivate = i;
+						break;
+					}
+				}
+
+			}
+		}
+		return base.OnInvokeKeyBindings (keyEvent);
+	}
+
+	// Recurse the child menus looking for a shortcut that matches the key
+	bool FindShortcutInChildMenu (Key key, MenuBarItem menuBarItem, out MenuItem menuItemToSelect)
+	{
+		menuItemToSelect = null;
+
+		if (menuBarItem?.Children == null) {
```

Comment:
Must prevents for `Key.Null.`

`
		if (key == Key.Null || menuBarItem?.Children == null) {
`